include/seastar/websocket/server.hh

```diff
@@ -39,19 +44,162 @@ struct reply {
+    }
+    uint8_t get_fin() {return fin;}
+    uint8_t get_rsv1() {return rsv1;}
+    uint8_t get_rsv2() {return rsv2;}
+    uint8_t get_rsv3() {return rsv3;}
+    uint8_t get_opcode() {return opcode;}
+    uint8_t get_masked() {return masked;}
+    uint8_t get_length() {return length;}
+
+    bool is_opcode_known() {
+        //https://datatracker.ietf.org/doc/html/rfc6455#section-5.1
+        return opcode < 0xA && !(opcode < 0x8 && opcode > 0x2);
+    }
+};
+
+
+
+class websocket_parser {
+    enum class parsing_state : uint8_t {
+        flags_and_payload_data,
+        payload_length_and_mask,
+        payload
+    };
+    enum class connection_state : uint8_t {
+        valid,
+        closed,
+        error
+    };
+    using consumption_result_t = consumption_result<char>;
+    using buff_t = temporary_buffer<char>;
+    // What parser is currently doing.
+    parsing_state _state;
+    // State of connection - can be valid, closed or should be closed 
+    // due to error.
+    connection_state _cstate;
+    sstring _buffer;
+    std::unique_ptr<frame_header> _header;
+    uint64_t _payload_length;
+    uint32_t _masking_key;
+    buff_t _result;
+
+    static future<consumption_result_t> dont_stop() {
+        return make_ready_future<consumption_result_t>(continue_consuming{});
+    }
+    static future<consumption_result_t> stop(buff_t data) {
+        return make_ready_future<consumption_result_t>(stop_consuming(std::move(data)));
+    }
+
+    // Removes mask from payload given in p.
+    void remove_mask(buff_t& p, size_t n) {
+        char *payload = p.get_write();
+        for (uint64_t i = 0, j = 0; i < n; ++i, j = (j + 1) % 4) {
+            payload[i] ^= static_cast<char>(((_masking_key << (j * 8)) >> 24));
+        }
+    } 
+public:
+    websocket_parser() : _state(parsing_state::flags_and_payload_data),
+                         _cstate(connection_state::valid),
+                         _payload_length(0), 
+                         _masking_key(0) {}
+    future<consumption_result_t> operator()(temporary_buffer<char> data);
+    bool is_valid() { return _cstate == connection_state::valid; }
+    bool eof() { return _cstate == connection_state::closed; }
+    buff_t result() { return std::move(_result); }
+};
+
 /*!
  * \brief a WebSocket connection
  */
 class connection : public boost::intrusive::list_base_hook<> {
+    using buff_t = temporary_buffer<char>;
+
+    /*!
+     * \brief Implementation of connection's data source.
+     */
+    class connection_source_impl final : public data_source_impl {
+        queue<buff_t>* data;
+
+    public:
+        connection_source_impl(queue<buff_t>* data) : data(data) {}
+
+        virtual future<buff_t> get() override {
+            return data->pop_eventually();
+        }
+
+        virtual future<> close() override {
+            data->abort(std::make_exception_ptr(exception("Connection closed")));
+            return make_ready_future<>();
```

Comment:
It's weird to have a close() returning a future, and yet the future doesn't wait for anything. For example, you mustn't destroy the "data" when close() returns, because there is another user of this queue (the one who will get that "Connection closed" exception) that can still read it.
Why do we need this close() at all?

Reply:
@nyh that's simply an implementation of a pure virtual function, required by the fact that it derives from `data_source_impl`. This websocket implementation exposes a custom stream of data, already decoded from the WebSocket protocol. It's mainly needed because WebSocket data is often masked with a simple xor key, and historical reasons for that are explained e.g. here: https://stackoverflow.com/questions/14174184/what-is-the-mask-in-a-websocket-frame

Reply:
I see.
Anyway, you're obviously much more experienced in this code than I am, so feel free to commit it when you feel it's ready.

Reply:
It's better to have close() wait for clean termination, and have a separate abort() to stop the connection. That's what all our other connection classes do.

Also, it /may/ be worthwhile to expose websockets as our regular socket hierarchy - so you can listen() and get a regular connected_socket out of it.